### PR TITLE
Fix #119

### LINKED ISSUE #119: [Bug]: AmpycloudError: Cloud base calculation got an empty array.Maybe check lookback percentage (is set to 100)
### Is there already an issue for this?

- [X] I found no existing issue for this bug.

### Describe the bug

A user reported the following error:
```
AmpycloudError: Cloud base calculation got an empty array.Maybe check lookback percentage (is set to 100)
```

This was tracked down to the `GaussianMixture` routine from sklearn, used in `src/ampycloud/layers.py`'s `ncomp_from_gmm()` routine. There are two sides to the problem. 

First, the line `with utils.tmp_seed(random_seed):` does not have the intended effect. That is, it does not seem to robustly fix the random_seed inside `GaussianMixture`. **Proposed fix: use `random_state=random_seed` as an additional argument of `GaussianMixture()` ?**

The second bug is more profound. Under certain circumstances, asking `GaussianMixture()` to look for `n` components will only assign `m<n` labels to the input values. This blows up afterwards when trying to find the base height of the sub-layers, because one of these will be comprised of no measurements.

A MWE to reproduce the problem is shown below. On my machine, this returns:
```
4 [0 2]
13 [0 1]
25 [0 1]
28 [0 1]
29 [0 1]
32 [0 2]
56 [0 2]
59 [0 1]
60 [0 2]
66 [0 2]
70 [0 1]
77 [0 1]
79 [1 2]
88 [0 1]
97 [0 2]
98 [0 1]
99 [0 1]
```

This implies that for a handful of random seeds, the values are assigned to only two sub-layers. And not always the same ones ! Why is that ? No idea, but I strongly suspect that this only happens when the layer does not have sub-structures.

In terms of a fix, one reasonable option would be to validate that each sub-layers has some points attributed to it. Should this not be the case, the `abics` score of this specific run gets boosted to be larger than all other options, which would be equivalent to say that this is a bad choice. This can be only a problem with 2 or 3 components, so we can always fall back on 1 component only.

### Screenshot

_No response_

### Steps to reproduce

```
import numpy as np
from sklearn.mixture import GaussianMixture

vals = np.array([[ 30.3030303 ],
       [ 18.18181818],
       [ 30.3030303 ],
       [ 18.18181818],
       [ 18.18181818],
       [  9.09090909],
       [  0.        ],
       [  0.        ],
       [ 30.3030303 ],
       [ 39.39393939],
       [ 39.39393939],
       [ 48.48484848],
       [ 48.48484848],
       [ 60.60606061],
       [ 39.39393939],
       [ 30.3030303 ],
       [ 30.3030303 ],
       [ 30.3030303 ],
       [ 48.48484848],
       [ 48.48484848],
       [ 30.3030303 ],
       [ 30.3030303 ],
       [ 39.39393939],
       [ 78.78787879],
       [100.        ],
       [ 48.48484848],
       [ 48.48484848],
       [ 69.6969697 ],
       [ 60.60606061],
       [ 69.6969697 ],
       [ 60.60606061],
       [ 60.60606061],
       [ 60.60606061],
       [ 48.48484848],
       [ 48.48484848],
       [ 60.60606061],
       [ 69.6969697 ],
       [ 87.87878788],
       [100.        ],
       [ 69.6969697 ],
       [ 60.60606061],
       [100.        ],
       [100.        ],
       [ 48.48484848],
       [ 48.48484848],
       [ 39.39393939],
       [ 30.3030303 ],
       [ 48.48484848],
       [ 39.39393939],
       [ 30.3030303 ],
       [ 39.39393939],
       [ 60.60606061],
       [ 69.6969697 ],
       [ 39.39393939],
       [  9.09090909],
       [  9.09090909],
       [ 30.3030303 ],
       [ 30.3030303 ]])

# Let's do the test for random seed values from 0 to 99
for state in range(100):
    # Optionally, run the test 10 times to make sure the outcome is always the same for a given seed
    #for i in range(10):

        # Run the GaussianMixture
        out = GaussianMixture(3, covariance_type='spherical', random_state=state).fit_predict(vals)
        # Count the number of labels identified. We expect to have hits falling in 3 sub-layers
        n_ids = len(np.unique(out))

        # If I have less than 3 sub-layers, print the corresponding random seed and ids
        if n_ids != 3:
            print(state, np.unique(out))
```

### Relevant log output

_No response_

### Environment

```markdown
OS X, ampycloud bleeding edge
```

### Code of Conduct

- [X] I agree to follow the ampycloud Code of Conduct.

### PR DESCRIPTION
**Pull request by fpavogt**
**Description:**

This PR fixes #119. It does so by:
 a) correctly feeding a `random_seed` to the `GaussianMixture()` call, and 
 b) boosts the `abics` score of any option fo which not all the sub-layers are populated as a result of the `.predict()` call.

**Error(s) fixed:**

Fixes #119.

**Checklists**:
- [ ] New code includes dedicated tests.
- [x] New code has been linted.
- [x] New code follows the project's style.
- [x] New code is compatible with the 3-Clause BSD license.
- [x] CHANGELOG has been updated.
- [ ] AUTHORS has been updated.
- [ ] Copyright years in module docstrings have been updated.

### CODE REVIEW

**Comment by fpavogt** [comment]
@regDaniel: I'll be curious to see what you think of this option. We can talk about it live.

**Review by regDaniel — APPROVED** [review]
Thanks for the fix @fpavogt ! I think it makes sense to artificially increase the score to reflect that we are dealing with a poor model.

File: src/ampycloud/layer.py
Comment (fpavogt) [reply src/ampycloud/layer.py:226]: Well, the layering step is definitely important. It is also robust. And in fact I think we did set the random seed in a "ok-ish" way, because there is a dedicated test for it (`def test_unstable_layers()` in `test_layers.py`). 

However, I think the way it was set before was not the most ideal, i.e. it would not carry over if one interrupts with a `pdb.set_trace()`. The new approach (feeding the random seed directly to `GaussianMixture()`) does work in that case.

File: src/ampycloud/layer.py
Comment (fpavogt) [reply src/ampycloud/layer.py:239]: I'm not in favor of a retry for the following two reasons: first, what if the retry also fails ? How many random seeds should one try before giving up ? I see a risk here to loop over many times to find a "proper" seed, which is not ideal. Second, as far as I understand, this situation occurs only when we try to fit 3 Gaussians to a distribution that obviously is NOT comprised of three Gaussians. There's no need to force `GaussianMixture()` to give us three "populated" Gaussians to decide if they make sense or not: the simple fact of not give 3 populated Gaussian tells us directly that the option makes no sense - boosting the score is just a way to rule it out (and since the score itself is not used for anything else).

**Comment by fpavogt** [comment]
@regDaniel: Thanks for the review. If ok for you, I'd propose to merge this into `develop_vof` once #116 has been reviewed, but before it gets merged. 

I'd propose to do that before merging the changes made by @srethore. What do you think ?

File: src/ampycloud/layer.py
Comment (regDaniel) [reply src/ampycloud/layer.py:239]: Fully agree.

**Comment by regDaniel** [comment]
Sounds good @fpavogt , I can then take care of rebasing @srethore 's branch before merging.

**Pull request merged**
